chain/find.go

```diff
@@ -56,231 +58,158 @@ func (g *GapIndexer) Run(ctx context.Context) error {
-		Where("height <= ?", g.maxHeight).
-		Select(); err != nil {
-		return nil, xerrors.Errorf("query processing report skips: %w", err)
-	}
-	log.Debugw("executed find skipped epoch query", "count", len(skippedReports))
+func (g *GapIndexer) findMissingTasksAndEpochs(ctx context.Context) (visor.GapReportList, error) {
+	log.Debug("finding task epoch gaps for heights", g.minHeight, "through", g.maxHeight)
+	start := time.Now()
 
-	gapReport := make([]*visor.GapReport, len(skippedReports))
-	for idx, r := range skippedReports {
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		default:
-		}
-		gapReport[idx] = &visor.GapReport{
-			Height:     r.Height,
-			Task:       r.Task,
-			Status:     "GAP",
-			Reporter:   g.name,
-			ReportedAt: reportTime,
-		}
+	var (
+		result           []TaskHeight
+		out              visor.GapReportList
+		sqlFmtTaskValues []string
+	)
+
+	for t := range g.taskSet.Iter() {
+		sqlFmtTaskValues = append(sqlFmtTaskValues, fmt.Sprintf("('%s')", t))
 	}
-	return gapReport, nil
-}
 
-func (g *GapIndexer) findEpochGapsAndNullRounds(ctx context.Context, node GapIndexerLens) (visor.GapReportList, []abi.ChainEpoch, error) {
-	log.Debug("finding epoch gaps and null rounds")
-	reportTime := time.Now()
+	// returns a list of tasks and heights for all incomplete heights and incomplete height
+	// is a height without an 'OK' or 'NULL_ROUND' for g.tasks. Returned values indicate
+	// heights and tasks which need to be filled.
+	query := fmt.Sprintf(`
+with
 
-	var nullRounds []abi.ChainEpoch
-	var missingHeights []uint64
-	res, err := g.DB.AsORM().QueryContext(
-		ctx,
-		&missingHeights,
-		`
-		SELECT s.i AS missing_epoch
-		FROM generate_series(?, ?) s(i)
-		WHERE NOT EXISTS (SELECT 1 FROM visor_processing_reports WHERE height = s.i);
-		`,
-		g.minHeight, g.maxHeight)
-	if err != nil {
-		return nil, nil, err
-	}
-	log.Debugw("executed find epoch gap query", "count", res.RowsReturned())
+-- generate all heights in range
+interesting_heights as (
+	select *
+	from generate_series(
+		?0,
+		?1
+	)
+	as x(height)
+)
+,
+
+-- enum all tasks for which we want to find gaps
+interesting_tasks as (
+	select *
+	from (values %s
+		-- example values in sqlFmtTasks:
+		-- ('actorstatesraw'),
+		-- ('actorstatespower'),
+		-- ('actorstatesreward'),
+		-- ('actorstatesminer'),
+		-- ('actorstatesinit'),
+		-- ('actorstatesmarket'),
+		-- ('actorstatesmultisig'),
+		-- ('actorstatesverifreg'),
+		-- ('blocks'),
+		-- ('messages'),
+		-- ('chaineconomics'),
+		-- ('msapprovals'),
+		-- ('implicitmessage'),
+		-- ('consensus')
+	) as x(task)
+)
+,
 
-	gapReport := make([]*visor.GapReport, 0, len(missingHeights))
-	// walk the possible gaps and query lotus to determine if gap was a null round or missed epoch.
-	for _, gap := range missingHeights {
-		select {
-		case <-ctx.Done():
-			return nil, nil, ctx.Err()
-		default:
-		}
-		gh := abi.ChainEpoch(gap)
-		tsgap, err := node.ChainGetTipSetByHeight(ctx, gh, types.EmptyTSK)
```

Comment:
This changes the behavior of gap find by causing it to label undetected null rounds as gaps. I think it will be better to address this as a separate change taking into consideration the conversation here: https://github.com/filecoin-project/lily/issues/767